language-plutus-core/src/Language/PlutusCore/TypeSynthesis.hs

```diff
@@ -220,47 +242,86 @@ typeOf (TyInst x body ty) = do
+rewriteCtx (TyApp x ty ty')     = TyApp x <$> rewriteCtx ty <*> rewriteCtx ty'
+rewriteCtx (TyFun x ty ty')     = TyFun x <$> rewriteCtx ty <*> rewriteCtx ty'
+rewriteCtx (TyFix x tn ty')     = TyFix x tn <$> rewriteCtx ty'
+rewriteCtx (TyLam x tn k ty)    = TyLam x tn k <$> rewriteCtx ty
+rewriteCtx (TyForall x tn k ty) = TyForall x tn k <$> rewriteCtx ty
+rewriteCtx ty@TyInt{}           = pure ty
+rewriteCtx ty@TyBuiltin{}       = pure ty
+rewriteCtx ty@(TyVar _ (TyNameWithKind (TyName (Name _ _ u)))) = do
+    (st, _) <- get
+    case IM.lookup (unUnique u) st of
+        Just ty'@(NormalizedType TyVar{}) -> rewriteCtx (getNormalizedType ty')
+        Just ty'                          -> cloneType (getNormalizedType ty')
+        Nothing                           -> pure ty
+
 -- | Reduce any redexes inside a type.
-tyReduce :: Type TyNameWithKind a -> NormalizedType TyNameWithKind a
--- TODO: is this case actually safe? Don't we need to reduce again after substituting?
-tyReduce (TyApp _ (TyLam _ (TyNameWithKind (TyName (Name _ _ u))) _ ty) ty') = NormalizedType $ tySubstitute u (NormalizedType ty') (tyReduce ty) -- TODO: use the substitution monad here
-tyReduce (TyForall x tn k ty)                                                = NormalizedType $ TyForall x tn k (getNormalizedType $ tyReduce ty)
-tyReduce (TyFun x ty ty') | isTypeValue ty                                   = NormalizedType $ TyFun x (getNormalizedType $ tyReduce ty) (getNormalizedType $ tyReduce ty')
-                          | otherwise                                        = NormalizedType $ TyFun x (getNormalizedType $ tyReduce ty) ty'
-tyReduce (TyLam x tn k ty)                                                   = NormalizedType $ TyLam x tn k (getNormalizedType $ tyReduce ty)
-tyReduce (TyApp x ty ty') | isTypeValue ty                                   = NormalizedType $ TyApp x (getNormalizedType $ tyReduce ty) (getNormalizedType $ tyReduce ty')
-                          | otherwise                                        = NormalizedType $ TyApp x (getNormalizedType $ tyReduce ty) ty'
-tyReduce x                                                                   = NormalizedType x
+tyReduce :: Type TyNameWithKind () -> TypeCheckM a (NormalizedType TyNameWithKind ())
+tyReduce (TyApp _ (TyLam _ (TyNameWithKind (TyName (Name _ _ u))) _ ty) ty') = do
+    tyEnvAssign u (NormalizedType (void ty'))
+    tyReduce ty <* tyEnvDelete u
+tyReduce (TyForall x tn k ty)                                                = NormalizedType <$> (TyForall x tn k <$> (getNormalizedType <$> tyReduce ty))
+tyReduce (TyFun x ty ty') | isTypeValue ty                                   = NormalizedType <$> (TyFun x <$> (getNormalizedType <$> tyReduce ty) <*> (getNormalizedType <$> tyReduce ty'))
+                          | otherwise                                        = NormalizedType <$> (TyFun x <$> (getNormalizedType <$> tyReduce ty) <*> rewriteCtx ty')
```

Comment:
I have no idea what happens here.

Reply:
This is necessary for spec compliance.

Reply:
When I implemented the CK machine from the spec, I added all the rules in comments, because this makes it much reader to understand the code and check its correctness against the specified rules. "Necessary for spec compliance" doesn't give the reader an idea of what's happening. This is not nitpicking, I really do not understand.